src/services/web-monitoring-db.js

```diff
@@ -193,18 +194,38 @@ export default class WebMonitoringDb {
   getPage (pageId) {
     return this._request(this._createUrl(`pages/${pageId}`))
       .then(response => response.json())
+      .then(throwIfError(`Could not load page: ${pageId}`))
       .then(data => parsePage(data.data));
   }
 
   /**
-     * Get a list of versions of a given page.
+     * Recursively get a list of all versions for a given page.
      * @param {string} pageId
+     * @param {string} query
      * @returns {Promise<Version[]>}
      */
-  getVersions (pageId) {
-    return this._request(this._createUrl(`pages/${pageId}/versions`))
-      .then(response => response.json())
-      .then(data => data.data.map(parseVersion));
+
+  getAllVersions (pageId, query) {
+    const api = this;
+    const url = api._createUrl(`pages/${pageId}/versions`, query);
+
+    return new Promise((resolve, reject) => {
+      let allVersions = [];
+      let getVersionsChunk = url => {
+        api._request(url)
+          .then(response => response.json())
+          .then(throwIfError(`Could not load versions from: ${url}`))
+          .then(data => {
+            allVersions.push(...data.data.map(parseVersion));
+            if (!data.links.next) {
+              resolve(allVersions);
+            } else {
+              getVersionsChunk(data.links.next);
+            }
+          });
+      };
+      getVersionsChunk(url);
```

Comment:
I think @Mr0grog is suggesting something like:

```es6
 * getVersions(pageId, query) {
      // a Generator method that yields one parsed Version at a time
      let url = api._createUrl(`pages/${pageId}/versions`, query);  // url of first page results
      while (true) {
        let data = this._request(url)
          .then(response => response.json());
        for (version in data.data) {
          yield parseVersion(version);
        }
        if (!data.links.next) {
          // No more pages of Versions to parse. The generator has been exhausted.
          break;
        }
        url = data.links.next;  // url of next page of results
      }
```

Apologies if that's pretty garbled ES6.

Reply:
For what it’s worth, that’s pretty close:

```js
// Gotta be an *async* generator, not a normal generator
async * getVersions(pageId, query) {
  let url = this._createUrl(`pages/${pageId}/versions`, query);
  while (url) {
    // The `await` here is key
    let data = await this._request(url)
      .then(response => response.json())
      .then(throwIfError(`Error getting ${url}`));
      
    for (version in data.data) {
      yield parseVersion(version);
    }
    
    url = data.links.next;
  }
}
```

But support for that is only shipping in Firefox and Chrome (and Safari Tech Preview, but that’s a pre-release thing), so maybe not something to rely on yet.